### ISSUE TITLE
Commit 887d43a Uses More RAM than it should/Wrong Batch size and Closing the Program Just Causes BSOD

### ISSUE THREAD

**Issue by erm14254** [p-e-w/heretic#263]
I have had 3 BSOD because of this issue, the auto detection uses a lot of RAM and then detects the biggest size possible, this is not an issue with PIQA, but its definitely an issue with both MMLU Pro and MMLU, the auto detection uses a lot of RAM due to testing for largest batch size and when it finishes it chooses the wrong batch size of 64 (too big), using CTRL+C doesn't work and closing the program causes BSOD.

The largest batch size for this test (MMLU) should be 32 (tested using my external lm-eval) for 24B model.

**Comment by p-e-w** [comment]
I can't reproduce this on any of the Linux systems I've tested it on. In fact, automatic batch size determination appears to work well, and finds a batch size very close to what I have determined manually to be the optimal one. I have tried this with both MMLU and MMLU Pro without issues.

Either way, this isn't our bug. The detection is performed by [lm-evaluation-harness](https://github.com/EleutherAI/lm-evaluation-harness) and you should see the same behavior if you simply run `lm_eval` from the command line with `--batch_size auto`, as it runs the exact same code. Please file an issue upstream if you can reproduce this.

It should also be noted that a BSOD as a result of userspace activity is something that should *never* happen, and is the job of the operating system to prevent. A regular program consuming too much RAM should result in the program being killed by the OS, not in the entire system crashing. So there may be a misconfiguration on your system related to this.

**Comment by erm14254** [comment]
There is no misconfiguration, BSOD from sudden ~222 GB memory release is a Windows kernel issue when too much memory is freed at once (CTRL+C doesn't work due to the "auto" batch size selection decide to select a batch file that is too large, which causes OOM and prevents CTRL+C from working, forcing me to suddenly kill the whole process, which causes ~222GB to suddenly being freed at once and Windows can not handle that), this issue only happens during the Heretic integration of lm-eval benchmark which uses a lot more RAM than the stand-alone lm-eval version, also using external stand-alone lm-eval benchmarks works no issue as long as I do not select "auto" for batch size, I set it to 16 or 32 depending on model's size and MMLU works no issues.

**Comment by p-e-w** [comment]
So running a benchmark from Heretic (which invokes `lm_eval` with `batch_size` set to `auto`) behaves differently than running `lm_eval` from CLI with `--batch_size auto`?

**Comment by erm14254** [comment]
> So running a benchmark from Heretic (which invokes `lm_eval` with `batch_size` set to `auto`) behaves differently than running `lm_eval` from CLI with `--batch_size auto`?

Yes, from Heretic it uses way more RAM and also the result scores are not the same compared to the standalone lm-eval external scores even when everything should be the same due to it's deterministic nature of the test (you run lm-eval standalone version twice, you will get the same results twice because it's deterministic, while if you run the same exact test on both the Heretic integration and stand-alone, you will get different results even when everything should be the same).

**Comment by p-e-w** [comment]
Could you please post the exact command line you use to invoke the `lm_eval` CLI? I assume that you use the `lm_eval` version installed by Heretic, to ensure it's the same one used internally.

> everything should be the same due to it's deterministic nature of the test

Benchmarks are actually not deterministic. Nothing is, on the GPU. It's almost impossible to get fully deterministic behavior with PyTorch, and results can depend on details like tensor distribution across devices, batch ordering, etc.

This is true even if you set a fixed seed, and use greedy sampling. Some PyTorch operations are even *fundamentally* non-deterministic depending on the hardware backend, and there's essentially nothing you can do about it.

**Comment by erm14254** [comment]
> Could you please post the exact command line you use to invoke the `lm_eval` CLI? I assume that you use the `lm_eval` version installed by Heretic, to ensure it's the same one used internally.

Sure, here:

PIQA

lm_eval --model hf --model_args pretrained=D:\Qwen3.5-27B --tasks piqa --device cuda:0 --batch_size 128

MMLU:

lm_eval --model hf --model_args pretrained=D:\Qwen3.5-27B --tasks mmlu --device cuda:0 --batch_size 16


> Benchmarks are actually not deterministic. Nothing is, on the GPU. It's almost impossible to get fully deterministic behavior with PyTorch, and results can depend on details like tensor distribution across devices, batch ordering, etc.

But I do get the same results, it is reproducible, the only time results change is if I change the batch size, but if I use the same batch size it's the same result, at least it was during my tests.

**Comment by p-e-w** [comment]
Those commands don't use `--batch_size auto`. Are you seeing the same problem with the CLI if you pass `--batch_size auto`? Because if the problem also occurs there, the bug is not in Heretic.

**Comment by erm14254** [comment]
> Those commands don't use `--batch_size auto`. Are you seeing the same problem with the CLI if you pass `--batch_size auto`? Because if the problem also occurs there, the bug is not in Heretic.

Yes, this is what I am saying, `--batch_size auto` causes issue because it tends to choose batch size that is too big, causing OOM, but it's even more severe in Heretic, where in Heretic ~222GB would be used for whatever reason, where with the stand-alone version ~102GB would be used instead.

**Comment by p-e-w** [comment]
I'm not sure what I'm supposed to do here.

As far as I can tell, this is an upstream bug, and I can't even reproduce it. Automatic batch size determination works fine in both Heretic and the `lm_eval` CLI for me. And considering that `lm_eval` is a very widely used piece of software, I imagine it works fine for most other people as well.

**Comment by erm14254** [comment]
> I'm not sure what I'm supposed to do here.
> 
> As far as I can tell, this is an upstream bug, and I can't even reproduce it. Automatic batch size determination works fine in both Heretic and the `lm_eval` CLI for me. And considering that `lm_eval` is a very widely used piece of software, I imagine it works fine for most other people as well.

Okay it's fine, you can close this issue report then, I will just use the stand-alone version (since the Heretic integration has an overuse of RAM issue as previously mentioned) and only select something like `--batch_size 16` for MMLU tests.

**Comment by p-e-w** [comment]
Ok.
